[BEAM-4366] Flipping tests fix

VaclavPlajt opened the pull request:
See the 7fcc8eb comment for the fix explanation.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [X] Write a pull request description that is detailed enough to understand:
   - [X] What the pull request does
   - [X] Why it does it
   - [X] How it does it
   - [X] Why this approach
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Run `./gradlew build` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

VaclavPlajt committed:
[BEAM-4366] Two sources of test flipps found and fixed.
First was wrong call to 'writer.commit();' in BeamWriteSink's method marked as '@Teardown'. It was moved to method marked '@FinishBundle'. Writters are now created per bundle.
Second cause of test flips was NPE when acessing carbage-collected key from WeakHashMap in ListDataSource. The ListDataSource was redesigned so the WeakHashMap is not used at all.

je-ik commented on sdks/java/extensions/euphoria/euphoria-core/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/client/util/Pair.java:
```diff
@@ -28,7 +29,7 @@
  */
 @Audience(Audience.Type.CLIENT)
-public final class Pair<K, V> {
+public final class Pair<K, V> implements Serializable {
```
I think this should be
```
public final class Pair<K extends Serializable, V extends Serializable> implements Serializable
```
otherwise the `implements Serializable` in invalid.

VaclavPlajt replied:
I see your point. Generally speaking what is a meaning of having `serializable` container when items within it are not. 

Trouble is that enforcing `Pair<K extends Serializable, V extends Serializable> ` brings up 70 compilation errors. Those may bring another `implements Serialization` enforcements, for example keys in `ReduceByKey` and `Join` operators, and some code rewrites to avoid it.

je-ik replied:
Hm, that is definitely not intended to force keys in R(S)BK to be Serializable. That would be a major breaking change and mostly use-less. Maybe we could make other change - the Serializable is needed only in certain situations. Maybe we could create private static subclass of `Pair`, which would extend it and be serializable (with restrictions to the type parameters) and add additional method
```
  public static <K extends Serializable, V extends Serializable> Pair<K, V> ofSerializable(K k, V v)
```
and use it wherever we need to force the object to be Serializable?

je-ik replied:
We would probably have to use custom serialization, because the parent Pair class of the derived class is not serializable.

je-ik replied:
Another option might be to turn the Pair into interface.

dmvk replied:
I see your point but I don't like the idea of forcing user to have serializable data. Beam's KV is also serializable and does not enforce K, V to be serializable.

dmvk replied:
I'd say only use case for this are tests and "Create" transformation, which will enforce user to make data serializable anyway

je-ik replied:
Yes, the problem is that Pair is used internally for keyed operations. This definitely means, that we don't want the `K`, `V` to be Serializable in all cases. Looks like I will have to learn how to live with this. :-)

VaclavPlajt replied:
I includes 7de0ebc as a thought experiment. It allows to have both serializable `Pair` with and without type parameters enforced to be serializable. Please review and comment.

je-ik reviewed (commented):
:+1: Looks good overall and is cleaner solution, but I think the Pair declaration needs attention. If we have `Pair`s that include non serializable fields, then this PR breaks them (and it cannot be used in sources). Another big argument in favor of forcing Serializable fields is `Fold` helper method, which can take `identity` (might be Pair), which would throw very unintuitive exceptions then.

dmvk commented on sdks/java/extensions/euphoria/euphoria-beam/build.gradle:
```diff
@@ -31,6 +31,7 @@ dependencies {
     shadow 'com.google.code.findbugs:annotations:3.0.1'
     testCompile project(':beam-sdks-java-extensions-euphoria-testing')
     testCompile project(':beam-runners-direct-java')
+    testCompile 'org.slf4j:slf4j-simple:1.7.25'
```
package versions should not be hardcoded

dmvk commented on sdks/java/extensions/euphoria/euphoria-operator-testkit/build.gradle:
```diff
@@ -26,6 +26,7 @@ dependencies {
     testCompile project(':beam-runners-direct-java')
     testCompile project(':beam-sdks-java-extensions-euphoria-testing')
     shadow 'com.google.code.findbugs:annotations:3.0.1'
+    testCompile 'org.slf4j:slf4j-simple:1.7.25'
```
hardcoded version

VaclavPlajt replied:
Fixed.

dmvk reviewed (changes requested):
LGTM 👍 Please change hardcoded dependency versions and it's good to merge ;) Good job.

VaclavPlajt committed:
[BEAM-4366] Dynamic dependency versions.

VaclavPlajt committed:
[BEAM-4366] Fully serializable `Pair` (including key and value) introduced.
No braking API changes where necessarry. Final `Pair` class was made abstract with two subclasses `SimplePair` and `SerializablePair`.

dmvk commented on sdks/java/extensions/euphoria/euphoria-core/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/client/util/Pair.java:
```diff
@@ -29,21 +29,27 @@
  */
 @Audience(Audience.Type.CLIENT)
-public final class Pair<K, V> implements Serializable {
+public abstract class Pair<K, V> implements Serializable{
```
How about moving Serializable interface implementation to SerializablePair?
